discovery_child_development/notebooks/patents/google/01_explore_data.py

```diff
@@ -0,0 +1,155 @@
+# ---
+# jupyter:
+#   jupytext:
+#     cell_metadata_filter: -all
+#     comment_magics: true
+#     formats: ipynb,py
+#     text_representation:
+#       extension: .py
+#       format_name: light
+#       format_version: '1.5'
+#       jupytext_version: 1.15.2
+#   kernelspec:
+#     display_name: discovery_child_development
+#     language: python
+#     name: python3
+# ---
+
+# # Explore patents data
+#
+# - Filter out patents where combinations of keywords don't appear in the same sentence
+# - Inspect themes by clustering the data
+# - Characterise the clusters by summarising the characteristic patents
+# - Visualise the patent data
+
+# +
+from discovery_child_development.getters import patents
+from discovery_child_development.utils import keywords as kw
+import discovery_child_development.utils.cluster_analysis_utils as cau
+import pandas as pd
+import numpy as np
+
+import importlib
+
+importlib.reload(patents)
+importlib.reload(kw)
+importlib.reload(cau)
+
+import altair as alt
+
+alt.data_transformers.disable_max_rows()
+from sentence_transformers import SentenceTransformer
+
+model = SentenceTransformer("all-MiniLM-L6-v2")
+
+SEED = 42
+
+UMAP_PARAMS = {
+    "n_components": 50,
+    "n_neighbors": 10,
+    "min_dist": 0.5,
+    "spread": 0.5,
+}
+
+# +
+# Load keywords
+keywords = patents.get_keywords_from_s3()
+
+# Load patent data
+data_raw_df = patents.get_patents_from_s3()
+
+data_df = (
+    data_raw_df
+    # Combine title and abstract
+    .assign(text=lambda df: df["title"] + ". " + df["abstract"])
```

Comment:
Yeah, it's just a simple hack because I noticed that patent titles usually don't have a full stop at the end. Might be useful for OpenAlex as well!

I guess you might have papers that already have some punctuation (eg, a paper with a question in the title) but I don't think that'd be an issue for downstream processing. Of course, you could also implement a more careful approach that checks for those cases (don't think we need that at the moment).